composer/callbacks/runtime_estimator.py

```diff
@@ -0,0 +1,178 @@
+            return
+
+        assert self.start_dur is not None
+        assert self.start_time is not None
+        if elapsed_dur > self.start_dur:
+            elapsed_time = time.time() - self.start_time
+            elapsed_time -= self.total_eval_wct  # Subtract time spent evaluating
+            print(
+                f'Elapsed time: {elapsed_time}, elapsed duration: {elapsed_dur}, checkpoint duration: {self.start_dur}')
+            rate = elapsed_time / (elapsed_dur - self.start_dur)
+            remaining_time = rate * (1 - elapsed_dur)
+
+            print(f'Batch end. train remaining_time: {remaining_time}')
+
+            # Add remaining time from each evaluator using known frequencies. We explicitly compute
+            # frequency instead of using time interpolation to avoid saw tooth pattern in estimates
+            for dataloader_label, eval_wcts in self.eval_wct_per_label.items():
+                # Discard first eval_wct if possible as it often slower due to dataset downloading
+                eval_wct_avg = None
+                num_evals_finished = len(eval_wcts)
+                if num_evals_finished > 1:
+                    eval_wct_avg = sum(eval_wcts[1:]) / (num_evals_finished - 1)
+                else:
+                    eval_wct_avg = sum(eval_wcts) / num_evals_finished
+                eval_rate = self.eval_frequency_per_label[dataloader_label]
+                print(
+                    f'\tdataloader_label: {dataloader_label}, eval_wct_avg: {eval_wct_avg}, eval_rate: {eval_rate}, num_evals_finished: {num_evals_finished}'
+                )
+                if eval_rate > 0:
+                    num_total_evals = 1 / eval_rate
+                    remaining_calls = num_total_evals - num_evals_finished
+                    remaining_time += eval_wct_avg * remaining_calls
+                    print(f'\tEval time: {eval_wct_avg * remaining_calls}')
+
+            print(f'\tRemaining time: {remaining_time}')
+            logger.log_metrics({'wall_clock/remaining_estimate': remaining_time})
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        # If eval is called before training starts, ignore it
+        if not self._enabled or self.start_time is None:
+            return
+        self.total_eval_wct += state.eval_timestamp.total_wct.total_seconds()
+        # state.dataloader_label should always be non-None unless user explicitly sets evaluator
+        # label to None, ignoring type hints
+        assert state.dataloader_label is not None, 'evaluator label must not be None'
+        if state.dataloader_label not in self.eval_wct_per_label:
+            self.eval_wct_per_label[state.dataloader_label] = []
+        self.eval_wct_per_label[state.dataloader_label].append(state.eval_timestamp.total_wct.total_seconds())
+        elapsed_fraction = self.get_elapsed_duration(state)
+        if elapsed_fraction is None:
+            warnings.warn(
```

Comment:
is this possible to hit? batch_start should set `self._enabled` to `False` if `max_duration` is `None`, so I think this would never be hit?